lib/common/socket.c

```diff
@@ -1470,3 +1470,112 @@ void h2o_sliding_counter_stop(h2o_sliding_counter_t *counter, uint64_t now)
     /* recalc average */
     counter->average = counter->prev.sum / (sizeof(counter->prev.slots) / sizeof(counter->prev.slots[0]));
 }
+
+#if H2O_USE_DTRACE && defined(__linux__)
+#include <linux/bpf.h>
+#include <linux/unistd.h>
+#include "h2o-probes.h"
+#include "include/h2o/ebpf.h"
+#include <sys/stat.h>
+
+int h2o_tracing_map_fd = 0;
+uint32_t h2o_tracing_map_last_attempt = 0;
+
+void open_map()
+{
+    // only check every second
+    uint32_t now = (uint32_t)time(NULL);
+    if (h2o_tracing_map_last_attempt == now)
+        return;
+
+    __sync_bool_compare_and_swap(&h2o_tracing_map_last_attempt, h2o_tracing_map_last_attempt, now);
+
+    // check if map exists at path
+    struct stat s;
+    if (stat(&H2O_EBPF_MAP_PATH[0], &s) == -1) {
+        // map path unavailable, cleanup fd if needed and leave
+        if (h2o_tracing_map_fd > 0) {
+            close(h2o_tracing_map_fd);
+            __sync_fetch_and_and(&h2o_tracing_map_fd, 0);
```

Comment:
I think we have a TOCTOU problem here.

We need to ensure that `h2o_tracing_map_fd` is never used (for example as an argument being passed to `syscall(__NR_bpf, ...)`) once close (2) is being called.

Otherwise, something like below can happen:
1. thread A checks that h2o_tracing_map_fd is valid, and builds bpf_attr using the value
2. thread B closes h2o_tracing_map_fd
3. thread C open a file (or socket or what ever), for which the same file descriptor is assigned
4. thread A calls `syscall(__NR_bpf)` using the bpf_attr being built in step 1.

As you can see, NR_bpf is receiving an open file description of something unexpected as its argument.

As a solution, I think the easiest way would be to make `h2o_tracing_map_fd` a thread-local variable. Another approach would be to convert `h2o_tracing_map_fd` to a refcounted object using atomic instructions.